Add pylint and bandit to tox and adapt code to pass linting

lukpueh opened the pull request:
**Description of the changes being introduced by the pull request**:
Adds a lint test environment to tox.ini to run bandit and pylint using securesystemslib's generic (pylintrc)[https://github.com/secure-systems-lab/sample-documents/blob/master/pylintrc].
We separate static from dynamic testing, because former needs to run only once (not for each tested Python version).

The PR also fixes various issues pointed out by running the linters.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Add pylint configuration file

The configuration file is taken as is from securesystemslib's
sample documents:
https://github.com/secure-systems-lab/sample-documents/blob/master/pylintrc

Credit goes to @vladimir-v-diaz

Commit:
Add pylint and bandit to tox.ini

Add new test environment that installs pylint and bandit
and executes them for static code analysis.
We separate linting from other test environments because it is
runtime version independent.

Commit:
Remove unused imports

The newly added linter pointed out obsolete import statements.
This commit removes them.

Commit:
Remove relative import found with pylint

Commit:
Remove/repurpose/rename unused variables

This commit cleans up variables that are never accessed (found
by pylint) appropriately.

Commit:
Remove duplicate docstring found by pylint

Commit:
Fix XXX comments disallowed by pylint

Commit:
Fix wrong indentation found by pylint

Commit:
Optimize logging-format-interpolation

Pylint complains about non-optimal string formatting in log
statements. This commit refactors the log module to pass
the string format and arguments to the logger instead of
formatting it beforehand.

https://docs.python.org/3/howto/logging.html#optimization

Commit:
Refactor protected variable access

Pylint points out that directly accessing a protected member
("_type") from outside the class (Metablock, Link, Layout) is a
bad idea.

This commit adds read-only properties to these classes and
adopts the calls accordingly.

Since a non-protected "type" variant of the attribute would
collide with the Python keyword, we apply the
single_trailing_underscore_ pep8 convention.

Commit:
Replace unsafe list default argument

Fix an occurrence of the mutable default argument problem.

Commit:
Fix unassigned variable bugs

Pylint found two bugs in the <1% code that is not unit tested.
Hooray for pylint, booh for missing unit tests!

vladimir-v-diaz commented:
GitHub indicates that coverage/coveralls failed.  Do you know why?  I took a quick glance but the coveralls output wasn't too helpful.

vladimir-v-diaz commented:
The Travis output is inundated with test statements.  You should consider printing only test failures to make it easier to review the output and/or any test failures.

lukpueh commented:
I think it is because the number of covered lines decreased in some files (I removed obsolete imports), which makes not covered lines weigh more and therefor decrease the overall coverage.

vladimir-v-diaz commented:
What happens if there is a Pylint or Bandit failure?  Does the travis build fail?  Is the pull request blocked if there are any linting failures?

vladimir-v-diaz commented:
> I think it is because the number of covered lines decreased in some files (I removed obsolete imports), which makes not covered lines weigh more and therefor decrease the overall coverage.

I wonder why GitHub flags the pull request as a failure because of the coveralls check.

vladimir-v-diaz commented on in_toto/models/layout.py:
```diff
@@ -101,6 +99,11 @@ def __init__(self, **kwargs):
+  @property
+  def type_(self):
+    """Getter for protected _type attribute. Trailing underscore used by
+    convention (pep8) to avoid conflict with Python's type keyword. """
```
More importantly, the trailing underscore convention is recommended by our code style guidelines.
https://github.com/secure-systems-lab/code-style-guidelines#function-and-method-arguments

lukpueh commented:
*Re verbosity:* I added a comment in a related issue and will fix it there https://github.com/in-toto/in-toto/issues/117#issuecomment-355584168.
*Re fail:* If pylint or bandit fail, they return a non-zero value which aborts tox `commands` execution and returns a non-zero exit code, which in turn makes the travis build fail.

vladimir-v-diaz commented:
> Re fail: If pylint or bandit fail, they return a non-zero value which aborts tox commands execution and returns a non-zero exit code, which in turn makes the travis build fail.

Another thing you should verify is whether the output of a pylint/bandit failure is printed to the console.  If there is a pylint or bandit failure, is it clear from reading Travis's output?

vladimir-v-diaz commented on in_toto/runlib.py:
```diff
@@ -263,7 +266,7 @@ def execute_link(link_cmd_args, record_streams):
 
   if record_streams:
-    # XXX: Use SpooledTemporaryFile if we expect very large outputs
+    # Use SpooledTemporaryFile if we expect very large outputs
```
Is this still a TODO?  It's not clear from the comment, now that `XXX:` is removed.

lukpueh replied:
Not really a todo, just a comment to consider this alternative if we run into problems.

vladimir-v-diaz replied:
"... to consider this alternative if we run into problems."

Maybe, "We should consider using SpooledTemporaryFile if we expect very large outputs in the future."

lukpueh commented:
*Re GitHub flag:* There are coveralls settings to configure thresholds, for minimum coverage and maximum coverage decrease.  I see for TUF you set 97% for the former. What to you think is a sensible default for the latter?

Commit:
Add tox lint environment to .travis.yml

Travis needs to be told which tox environments to run. This
commit adds the newly added lint tox environment to travis.

lukpueh commented:
> Another thing you should verify is whether the output of a pylint/bandit failure is printed to the console. If there is a pylint or bandit failure, is it clear from reading Travis's output?

Good catch, I just saw that travis actually didn't run the linters. This is fixed with 3a28ac9.

vladimir-v-diaz commented:
I'm not sure what's a good number for the maximum coverage decrease.  It is ideal that the majority of pull requests contain some unit testing, and that the minimum test coverage not fall below a certain threshold.

Is setting a minimum threshold below your current coverage sufficient?  Say, 99% minimum test coverage when your current coverage is 100%.  In this way, you can allow a large maximum coverage decrease to avoid any failed coveralls check, and still allow some pull requests that are not fully tested (perhaps it cannot be fully tested because it depends on another pending pull request).

vladimir-v-diaz commented on in_toto/models/link.py:
```diff
@@ -19,7 +19,6 @@
 import attr
-import json
 import securesystemslib.formats
 from in_toto.models.common import Signable
```
I see that the "from foo import bar" convention is being used.  Is there a reason why?  This convention is not advised by our guidelines

lukpueh replied:
This is not related to this PR. But acknowledged in https://github.com/in-toto/in-toto/issues/145.

lukpueh commented:
Good thoughts. I set the min coverage threshold to 99% and the max decrease to 0.5%. Latter should ignore fails due to line number decrease. 

In any case, we do allow and have performed merges that fail the coverage/coveralls test.

vladimir-v-diaz commented on tox.ini:
```diff
@@ -4,9 +4,26 @@
+    bandit
+
+commands =
+    # Run pylint, using securesystemslib's pylintrc configuration file
```
pylintrc is provided by "secure systems lab," not "securesystemslib."

vladimir-v-diaz commented on tox.ini:
```diff
@@ -4,9 +4,26 @@
+    # subprocess32 modules, which are considered to have low severity security
+    # implications, but are required by in-toto. We do not skip subprocess
+    # calls with shell=True (B602).
+    bandit -r in_toto --skip B404,B603
```
Can you link to B404 and B603, to make it easier for the reader to understand the implications of not testing them?

https://docs.openstack.org/bandit/latest/plugins/subprocess_without_shell_equals_true.html

https://docs.openstack.org/bandit/latest/plugins/subprocess_popen_with_shell_equals_true.html

https://docs.openstack.org/bandit/latest/blacklists/blacklist_imports.html#b404-import-subprocess

vladimir-v-diaz commented:
Okay, that's my final comments for the review.

Commit:
Update comment about SpooledTemporaryFile

Commit:
Add links to skipped bandit test in tox.ini

lukpueh commented:
Thanks for the review, @vladimir-v-diaz!
